pyhole/plugins/launchpad.py

```diff
@@ -14,96 +14,190 @@
 
 class Launchpad(plugin.Plugin):
     """Provide access to the Launchpad API."""
 
     def __init__(self, session):
         self.session = session
+        self.nick_cache = {}
+        # {'nick': 'launchpad-username',...}
+        self.nick_cache_path = CONF.get("nick_cache")
         self.name = self.__class__.__name__
-        self.launchpad = LP.login_anonymously("pyhole", "production",
-                                              utils.get_directory(self.name))
+        
+        self.auth_app_name = "{nick}@{app}".format(nick=self.session.nick,
+                                                   app=CONF.get("app_name"))
+        self.service_root = CONF.get("service_root")
+        self.target = CONF.get("target")
+        self.delay_between = CONF.get("delay_between", type="int")
+        self.contact = CONF.get("contact")  # whos being represented by the bot
+        self.credentials = Credentials(self.auth_app_name)  # "bot@pyhole"
+        self._authed = False
+    
+    @property
+    def authed(self):
+        return self._authed
 
-    @plugin.hook_add_command("lbugs")
-    @utils.require_params
+    @plugin.hook_add_command("lauth")
+    @utils.admin
     @utils.spawn
-    def lbugs(self, message, params=None, **kwargs):
-        """Launchpad bugs for a team (ex: .lbugs <project> <team>|<user>)."""
-        project, team = params.split(" ", 2)
+    def authenticate(self, message, params=None, **kwargs):
+        self._authed = False
         try:
-            members = self.launchpad.people[team]
-            proj = self.launchpad.projects[project]
+            auth_delay = CONF.get("auth_delay", type="int")
+            req_token_info = self.credentials.get_request_token(
+                web_root='production')
+            msg = ('Authenticate {app} as READ ANYTHING at: {token_info} '
+                    'within {delay} seconds'.format(app=self.auth_app_name,
+                                                    token_info=req_token_info,
+                                                    delay=auth_delay))
+            self.session.privmsg(self.contact, msg)
+            # Get back a string that looks like 
+            # 'https://launchpad.net/+authorize-token?oauth_token=...'
+            # This is the URL your end-user needs to visit in order to
+            # authorize your token.
+            # Give the user a moment to authorize the application
+            time.sleep(auth_delay)
+            try:
+                self.credentials.exchange_request_token_for_access_token(
+                    web_root=self.target)
+                self.launchpad = LP(self.credentials, None, None,
+                                    service_root='production')
+                self.session.privmsg(self.contact, "Launchpad authenticated.")
+                self._authed = True
+            except HTTPError:
+                # user hasn't authorized the token yet.
+                msg = "Not approved in the alloted time of %s seconds."
```

Comment:
I wanted to leave in the ability to login noauth, because there's still stuff you can do with that. I'm still a little unsure of what failure really means here.